File: server/billing/service.ts
import {
  planCheckoutCreation,
  planReconciliationSnapshot,
  planVerifiedProviderFact,
  toBillingSubscriptionFacts,
} from './core';
import type { BillingOwnershipPort, BillingRepository } from './ports';
import type {
  BillingApi,
  BillingCommandResult,
  ProviderFactResult,
} from './public';

export function createBillingApi(input: {
  readonly ownership: BillingOwnershipPort;
  readonly repository: BillingRepository;
}): BillingApi {
  return {
    async beginCheckout(context, command): Promise<BillingCommandResult> {
      if (!(await input.ownership.owns(context))) {
        return { kind: 'rejected', reason: 'owner-mismatch' };
      }
      const plan = planCheckoutCreation(context, command);
      if (plan.kind === 'rejected') return plan;
      const intent = {
        checkoutIntentId: command.checkoutIntentId,
        subscriptionId: command.subscriptionId,
        provider: command.provider,
        providerCheckoutReference: null,
        status: 'created',
        createdAt: command.createdAt,
        openedAt: null,
      } as const;
      const result = await input.repository.createCheckout({
        record: plan.record,
        intent,
      });
      if (result.kind === 'created') {
        return {
          kind: 'applied',
          facts: toBillingSubscriptionFacts(plan.record),
        };
      }
      if (
        result.record !== undefined &&
        result.intent !== undefined &&
        sameCheckout(result.record, result.intent, plan.record, intent)
      ) {
        return {
          kind: 'replayed',
          facts: toBillingSubscriptionFacts(result.record),
        };
      }
      return { kind: 'rejected', reason: 'identifier-conflict' };
    },

    async recordCheckoutOpened(context, command) {
      if (!(await input.ownership.owns(context))) {
        return { kind: 'rejected', reason: 'owner-mismatch' };
      }
      const [record, existing] = await Promise.all([
        input.repository.findByOwner(context),
        input.repository.findCheckoutIntent(command.checkoutIntentId),
      ]);
      if (
        record === undefined ||
        existing === undefined ||
        record.subscriptionId !== command.subscriptionId ||
        existing.subscriptionId !== command.subscriptionId
      ) {
        return { kind: 'rejected', reason: 'not-found' };
      }
      const mappedCheckout =
        await input.repository.findCheckoutByProviderReference(
          existing.provider,
          command.providerCheckoutReference,
        );
      if (
        mappedCheckout !== undefined &&
        mappedCheckout.checkoutIntentId !== command.checkoutIntentId
      ) {
        return { kind: 'rejected', reason: 'identifier-conflict' };
      }
      if (
        !Number.isSafeInteger(command.openedAt) ||
        command.openedAt < existing.createdAt
      ) {
        return { kind: 'rejected', reason: 'invalid-transition' };
      }
      const result = await input.repository.openCheckout({
        context,
        intent: {
          ...existing,
          providerCheckoutReference: command.providerCheckoutReference,
          status: 'opened',
          openedAt: command.openedAt,
        },
      });
      if (result.kind === 'conflict') {
        return { kind: 'rejected', reason: 'identifier-conflict' };
      }
      return {
        kind: result.kind === 'applied' ? 'applied' : 'replayed',
        facts: toBillingSubscriptionFacts(record),
      };
    },

    async ingestVerifiedProviderFact(fact): Promise<ProviderFactResult> {
      const [current, mapped] = await Promise.all([
        input.repository.findById(fact.subscriptionId),
        input.repository.findByProviderMapping({
          provider: fact.provider,
          customerReference: fact.providerCustomerReference,
          subscriptionReference: fact.providerSubscriptionReference,
        }),
      ]);
      if (current === undefined) {
        return { kind: 'rejected', reason: 'not-found' };
      }
      if (
        mapped !== undefined &&
        mapped.subscriptionId !== fact.subscriptionId
      ) {
        return { kind: 'rejected', reason: 'mapping-mismatch' };
      }
      const existing = await input.repository.findProviderEventReceipt(
        fact.provider,
        fact.eventId,
      );
      if (existing !== undefined) {
        return existing.subscriptionId === current.subscriptionId
          ? { kind: 'duplicate', facts: toBillingSubscriptionFacts(current) }
          : { kind: 'rejected', reason: 'mapping-mismatch' };
      }
      const plan = planVerifiedProviderFact(current, fact);
      if (plan.kind === 'rejected') return plan;
      const receipt = {
        provider: fact.provider,
        eventId: fact.eventId,
        subscriptionId: fact.subscriptionId,
        factKind: fact.kind,
        outcome: plan.kind === 'apply' ? 'applied' : 'ignored',
        occurredAt: fact.occurredAt,
        appliedVersion: plan.record.version,
        recordedAt: fact.recordedAt,
      } as const;
      if (plan.kind === 'ignore') {
        const stored =
          await input.repository.recordIgnoredProviderFact(receipt);
        return stored.kind === 'duplicate'
          ? { kind: 'duplicate', facts: toBillingSubscriptionFacts(current) }
          : {
              kind: 'ignored',
              reason: plan.reason,
              facts: toBillingSubscriptionFacts(plan.record),
            };
      }
      const committed = await input.repository.commitProviderFact({
        expectedRecord: current,
        nextRecord: plan.record,
        receipt,
      });
      return commitResult(
        committed,
        plan.record,
        async () =>
          (await input.repository.findById(fact.subscriptionId)) ?? current,
      );
    },

    async reconcileVerifiedSnapshot(snapshot): Promise<ProviderFactResult> {
      const [current, mapped] = await Promise.all([
        input.repository.findById(snapshot.subscriptionId),
        input.repository.findByProviderMapping({
          provider: snapshot.provider,
          customerReference: snapshot.providerCustomerReference,
          subscriptionReference: snapshot.providerSubscriptionReference,
        }),
      ]);
      if (current === undefined) {
        return { kind: 'rejected', reason: 'not-found' };
      }
      if (
        mapped !== undefined &&
        mapped.subscriptionId !== snapshot.subscriptionId
      ) {
        return { kind: 'rejected', reason: 'mapping-mismatch' };
      }
      const existing = await input.repository.findReconciliationCheckpoint(
        snapshot.provider,
        snapshot.snapshotId,
      );
      if (existing !== undefined) {
        return existing.subscriptionId === current.subscriptionId
          ? { kind: 'duplicate', facts: toBillingSubscriptionFacts(current) }
          : { kind: 'rejected', reason: 'mapping-mismatch' };
      }
      const plan = planReconciliationSnapshot(current, snapshot);
      if (plan.kind === 'rejected') return plan;
      const checkpoint = {
        provider: snapshot.provider,
        snapshotId: snapshot.snapshotId,
        subscriptionId: snapshot.subscriptionId,
        observedAt: snapshot.observedAt,
        appliedVersion: plan.record.version,
        recordedAt: snapshot.recordedAt,
      };
      if (plan.kind === 'ignore') {
        const stored =
          await input.repository.recordIgnoredReconciliation(checkpoint);
        return stored.kind === 'duplicate'
          ? { kind: 'duplicate', facts: toBillingSubscriptionFacts(current) }
          : {
              kind: 'ignored',
              reason: plan.reason,
              facts: toBillingSubscriptionFacts(plan.record),
            };
      }
      const committed = await input.repository.commitReconciliation({
        expectedRecord: current,
        nextRecord: plan.record,
        checkpoint,
      });
      return commitResult(
        committed,
        plan.record,
        async () =>
          (await input.repository.findById(snapshot.subscriptionId)) ?? current,
      );
    },

    async readSubscription(context) {
      if (!(await input.ownership.owns(context))) return undefined;
      const record = await input.repository.findByOwner(context);
      return record === undefined
        ? undefined
        : toBillingSubscriptionFacts(record);
    },
  };
}

function sameCheckout(
  existingRecord: Parameters<typeof toBillingSubscriptionFacts>[0],
  existingIntent: Parameters<BillingRepository['createCheckout']>[0]['intent'],
  proposedRecord: Parameters<typeof toBillingSubscriptionFacts>[0],
  proposedIntent: Parameters<BillingRepository['createCheckout']>[0]['intent'],
): boolean {
  return (
    existingRecord.subscriptionId === proposedRecord.subscriptionId &&
    existingRecord.accountId === proposedRecord.accountId &&
    existingRecord.vaultId === proposedRecord.vaultId &&
    existingRecord.provider === proposedRecord.provider &&
    existingIntent.checkoutIntentId === proposedIntent.checkoutIntentId &&
    existingIntent.subscriptionId === proposedIntent.subscriptionId &&
    existingIntent.provider === proposedIntent.provider &&
    existingIntent.createdAt === proposedIntent.createdAt
  );
}

async function commitResult(
  result: Awaited<ReturnType<BillingRepository['commitProviderFact']>>,
  planned: Parameters<typeof toBillingSubscriptionFacts>[0],
  readCurrent: () => Promise<Parameters<typeof toBillingSubscriptionFacts>[0]>,
): Promise<ProviderFactResult> {
  switch (result.kind) {
    case 'applied':
      return { kind: 'applied', facts: toBillingSubscriptionFacts(planned) };
    case 'duplicate':
      return {
        kind: 'duplicate',
        facts: toBillingSubscriptionFacts(await readCurrent()),
      };
    case 'conflict':
      return { kind: 'rejected', reason: 'cas-conflict' };
  }
}

File: backend/internal/billing/service.go
package billing

import (
	"context"
	"errors"

	"github.com/fukamu/notes/backend/internal/identity"
)

var ErrInvalidServiceConfiguration = errors.New("invalid billing service configuration")

type CommandResultKind string

const (
	ResultApplied   CommandResultKind = "applied"
	ResultReplayed  CommandResultKind = "replayed"
	ResultDuplicate CommandResultKind = "duplicate"
	ResultIgnored   CommandResultKind = "ignored"
	ResultRejected  CommandResultKind = "rejected"
)

type ResultReason string

const (
	ResultOwnerMismatch      ResultReason = "owner-mismatch"
	ResultIdentifierConflict ResultReason = "identifier-conflict"
	ResultNotFound           ResultReason = "not-found"
	ResultInvalidTransition  ResultReason = "invalid-transition"
	ResultMappingMismatch    ResultReason = "mapping-mismatch"
	ResultProviderMismatch   ResultReason = "provider-mismatch"
	ResultCASConflict        ResultReason = "cas-conflict"
	ResultStale              ResultReason = "stale"
	ResultTerminal           ResultReason = "terminal"
	ResultNoChange           ResultReason = "no-change"
)

type SubscriptionFacts struct {
	SubscriptionID     SubscriptionID
	AccountID          identity.AccountID
	VaultID            identity.VaultID
	Version            Version
	Lifecycle          Lifecycle
	PaymentMethodReady bool
	CancelAt           *int64
	UpdatedAt          int64
}

type CommandResult struct {
	Kind   CommandResultKind
	Reason ResultReason
	Facts  *SubscriptionFacts
}

type RecordCheckoutOpenedCommand struct {
	SubscriptionID            SubscriptionID
	CheckoutIntentID          CheckoutIntentID
	ProviderCheckoutReference ProviderCheckoutReference
	OpenedAt                  int64
}

type Service struct {
	ownership  OwnershipPort
	repository BillingRepository
}

func NewService(ownership OwnershipPort, repository BillingRepository) (*Service, error) {
	if ownership == nil || repository == nil {
		return nil, ErrInvalidServiceConfiguration
	}
	return &Service{ownership: ownership, repository: repository}, nil
}

func (service *Service) BeginCheckout(
	ctx context.Context,
	context identity.VaultContext,
	command BeginCheckoutCommand,
) (CommandResult, error) {
	owned, err := service.ownership.Owns(ctx, context)
	if err != nil {
		return CommandResult{}, err
	}
	if !owned {
		return rejectedResult(ResultOwnerMismatch), nil
	}
	plan := PlanCheckoutCreation(OwnerScope{AccountID: context.AccountID, VaultID: context.VaultID}, command)
	if plan.Kind == CheckoutPlanRejected {
		return rejectedResult(ResultInvalidTransition), nil
	}
	intent := CheckoutIntentRecord{
		CheckoutIntentID: command.CheckoutID,
		SubscriptionID:   command.SubscriptionID,
		Provider:         command.Provider,
		Status:           CheckoutIntentCreated,
		CreatedAt:        command.CreatedAt,
	}
	created, err := service.repository.CreateCheckout(ctx, plan.Record, intent)
	if err != nil {
		return CommandResult{}, err
	}
	if created.Kind == CheckoutCreated {
		return resultWithFacts(ResultApplied, "", plan.Record), nil
	}
	if created.Record != nil && created.Intent != nil && sameCheckout(*created.Record, *created.Intent, plan.Record, intent) {
		return resultWithFacts(ResultReplayed, "", *created.Record), nil
	}
	return rejectedResult(ResultIdentifierConflict), nil
}

func (service *Service) RecordCheckoutOpened(
	ctx context.Context,
	context identity.VaultContext,
	command RecordCheckoutOpenedCommand,
) (CommandResult, error) {
	owned, err := service.ownership.Owns(ctx, context)
	if err != nil {
		return CommandResult{}, err
	}
	if !owned {
		return rejectedResult(ResultOwnerMismatch), nil
	}
	current, err := service.repository.FindByOwner(ctx, OwnerScope{AccountID: context.AccountID, VaultID: context.VaultID})
	if err != nil {
		return CommandResult{}, err
	}
	intent, err := service.repository.FindCheckoutIntent(ctx, command.CheckoutIntentID)
	if err != nil {
		return CommandResult{}, err
	}
	if current == nil || intent == nil || current.SubscriptionID != command.SubscriptionID || intent.SubscriptionID != command.SubscriptionID {
		return rejectedResult(ResultNotFound), nil
	}
	if _, err := ParseProviderCheckoutReference(string(command.ProviderCheckoutReference)); err != nil ||
		!validTimestamp(command.OpenedAt) || command.OpenedAt < intent.CreatedAt {
		return rejectedResult(ResultInvalidTransition), nil
	}
	mapped, err := service.repository.FindCheckoutByProviderReference(ctx, intent.Provider, command.ProviderCheckoutReference)
	if err != nil {
		return CommandResult{}, err
	}
	if mapped != nil && mapped.CheckoutIntentID != command.CheckoutIntentID {
		return rejectedResult(ResultIdentifierConflict), nil
	}
	opened := *intent
	opened.ProviderCheckoutReference = command.ProviderCheckoutReference
	opened.Status = CheckoutIntentOpened
	opened.OpenedAt = pointer(command.OpenedAt)
	committed, err := service.repository.OpenCheckout(
		ctx,
		OwnerScope{AccountID: context.AccountID, VaultID: context.VaultID},
		opened,
	)
	if err != nil {
		return CommandResult{}, err
	}
	if committed == CommitConflict {
		return rejectedResult(ResultIdentifierConflict), nil
	}
	if committed == CommitApplied {
		return resultWithFacts(ResultApplied, "", *current), nil
	}
	return resultWithFacts(ResultReplayed, "", *current), nil
}

func (service *Service) IngestVerifiedProviderFact(
	ctx context.Context,
	fact VerifiedProviderFact,
) (CommandResult, error) {
	current, err := service.repository.FindByID(ctx, fact.SubscriptionID)
	if err != nil {
		return CommandResult{}, err
	}
	if current == nil {
		return rejectedResult(ResultNotFound), nil
	}
	mapped, err := service.repository.FindByProviderMapping(
		ctx, fact.Provider, fact.ProviderCustomerReference, fact.ProviderSubscriptionReference,
	)
	if err != nil {
		return CommandResult{}, err
	}
	if mapped != nil && mapped.SubscriptionID != fact.SubscriptionID {
		return rejectedResult(ResultMappingMismatch), nil
	}
	existing, err := service.repository.FindProviderEventReceipt(ctx, fact.Provider, fact.EventID)
	if err != nil {
		return CommandResult{}, err
	}
	if existing != nil {
		if existing.SubscriptionID != current.SubscriptionID {
			return rejectedResult(ResultMappingMismatch), nil
		}
		return resultWithFacts(ResultDuplicate, "", *current), nil
	}
	plan := PlanVerifiedProviderFact(*current, fact)
	if plan.Kind == ProviderFactRejected {
		return rejectedResult(mapPlanReason(plan.Reason)), nil
	}
	receipt := ProviderEventReceipt{
		Provider: fact.Provider, EventID: fact.EventID, SubscriptionID: fact.SubscriptionID,
		FactKind: fact.Kind, OccurredAt: fact.OccurredAt, AppliedVersion: plan.Record.Version, RecordedAt: fact.RecordedAt,
	}
	if plan.Kind == ProviderFactIgnore {
		receipt.Outcome = ReceiptIgnored
		stored, storeErr := service.repository.RecordIgnoredProviderFact(ctx, receipt)
		if storeErr != nil {
			return CommandResult{}, storeErr
		}
		if stored == CommitDuplicate {
			return resultWithFacts(ResultDuplicate, "", *current), nil
		}
		if stored == CommitConflict {
			return rejectedResult(ResultMappingMismatch), nil
		}
		return resultWithFacts(ResultIgnored, mapPlanReason(plan.Reason), plan.Record), nil
	}
	receipt.Outcome = ReceiptApplied
	committed, err := service.repository.CommitProviderFact(ctx, *current, plan.Record, receipt)
	if err != nil {
		return CommandResult{}, err
	}
	return service.commitResult(ctx, committed, plan.Record, fact.SubscriptionID)
}

func (service *Service) ReconcileVerifiedSnapshot(
	ctx context.Context,
	snapshot ReconciliationSnapshot,
) (CommandResult, error) {
	current, err := service.repository.FindByID(ctx, snapshot.SubscriptionID)
	if err != nil {
		return CommandResult{}, err
	}
	if current == nil {
		return rejectedResult(ResultNotFound), nil
	}
	mapped, err := service.repository.FindByProviderMapping(
		ctx, snapshot.Provider, snapshot.ProviderCustomerReference, snapshot.ProviderSubscriptionReference,
	)
	if err != nil {
		return CommandResult{}, err
	}
	if mapped != nil && mapped.SubscriptionID != snapshot.SubscriptionID {
		return rejectedResult(ResultMappingMismatch), nil
	}
	existing, err := service.repository.FindReconciliationCheckpoint(ctx, snapshot.Provider, snapshot.SnapshotID)
	if err != nil {
		return CommandResult{}, err
	}
	if existing != nil {
		if existing.SubscriptionID != current.SubscriptionID {
			return rejectedResult(ResultMappingMismatch), nil
		}
		return resultWithFacts(ResultDuplicate, "", *current), nil
	}
	plan := PlanReconciliationSnapshot(*current, snapshot)
	if plan.Kind == ProviderFactRejected {
		return rejectedResult(mapPlanReason(plan.Reason)), nil
	}
	checkpoint := ReconciliationCheckpoint{
		Provider: snapshot.Provider, SnapshotID: snapshot.SnapshotID, SubscriptionID: snapshot.SubscriptionID,
		ObservedAt: snapshot.ObservedAt, AppliedVersion: plan.Record.Version, RecordedAt: snapshot.RecordedAt,
	}
	if plan.Kind == ProviderFactIgnore {
		stored, storeErr := service.repository.RecordIgnoredReconciliation(ctx, checkpoint)
		if storeErr != nil {
			return CommandResult{}, storeErr
		}
		if stored == CommitDuplicate {
			return resultWithFacts(ResultDuplicate, "", *current), nil
		}
		if stored == CommitConflict {
			return rejectedResult(ResultMappingMismatch), nil
		}
		return resultWithFacts(ResultIgnored, mapPlanReason(plan.Reason), plan.Record), nil
	}
	committed, err := service.repository.CommitReconciliation(ctx, *current, plan.Record, checkpoint)
	if err != nil {
		return CommandResult{}, err
	}
	return service.commitResult(ctx, committed, plan.Record, snapshot.SubscriptionID)
}

func (service *Service) ReadSubscription(
	ctx context.Context,
	context identity.VaultContext,
) (*SubscriptionFacts, error) {
	owned, err := service.ownership.Owns(ctx, context)
	if err != nil || !owned {
		return nil, err
	}
	record, err := service.repository.FindByOwner(ctx, OwnerScope{AccountID: context.AccountID, VaultID: context.VaultID})
	if err != nil || record == nil {
		return nil, err
	}
	facts := factsFrom(*record)
	return &facts, nil
}

func (service *Service) commitResult(
	ctx context.Context,
	kind CommitKind,
	planned SubscriptionRecord,
	subscriptionID SubscriptionID,
) (CommandResult, error) {
	switch kind {
	case CommitApplied:
		return resultWithFacts(ResultApplied, "", planned), nil
	case CommitDuplicate:
		current, err := service.repository.FindByID(ctx, subscriptionID)
		if err != nil {
			return CommandResult{}, err
		}
		if current == nil {
			return rejectedResult(ResultNotFound), nil
		}
		return resultWithFacts(ResultDuplicate, "", *current), nil
	default:
		return rejectedResult(ResultCASConflict), nil
	}
}

func sameCheckout(
	existingRecord SubscriptionRecord,
	existingIntent CheckoutIntentRecord,
	proposedRecord SubscriptionRecord,
	proposedIntent CheckoutIntentRecord,
) bool {
	return existingRecord.SubscriptionID == proposedRecord.SubscriptionID &&
		existingRecord.AccountID == proposedRecord.AccountID && existingRecord.VaultID == proposedRecord.VaultID &&
		existingRecord.Provider == proposedRecord.Provider &&
		existingIntent.CheckoutIntentID == proposedIntent.CheckoutIntentID &&
		existingIntent.SubscriptionID == proposedIntent.SubscriptionID && existingIntent.Provider == proposedIntent.Provider &&
		existingIntent.CreatedAt == proposedIntent.CreatedAt
}

func factsFrom(record SubscriptionRecord) SubscriptionFacts {
	return SubscriptionFacts{
		SubscriptionID: record.SubscriptionID, AccountID: record.AccountID, VaultID: record.VaultID,
		Version: record.Version, Lifecycle: record.Lifecycle, PaymentMethodReady: record.PaymentMethodReady,
		CancelAt: cloneTimestamp(record.CancelAt), UpdatedAt: record.UpdatedAt,
	}
}

func resultWithFacts(kind CommandResultKind, reason ResultReason, record SubscriptionRecord) CommandResult {
	facts := factsFrom(record)
	return CommandResult{Kind: kind, Reason: reason, Facts: &facts}
}

func rejectedResult(reason ResultReason) CommandResult {
	return CommandResult{Kind: ResultRejected, Reason: reason}
}

func mapPlanReason(reason PlanReason) ResultReason {
	switch reason {
	case ReasonProviderMismatch:
		return ResultProviderMismatch
	case ReasonMappingMismatch:
		return ResultMappingMismatch
	case ReasonStale:
		return ResultStale
	case ReasonTerminal:
		return ResultTerminal
	case ReasonNoChange:
		return ResultNoChange
	default:
		return ResultInvalidTransition
	}
}
